### PR TITLE
feat(EMS-2816-2901-2902): No PDF - Policy - Loss policy financial details International - Data saving and encryption

### PR DESCRIPTION
**Pull request by Zainzzkk**
## Introduction :pencil2:
This PR adds data saving and encryption for Loss policy financial details international

## Resolution :heavy_check_mark:
* Used existing encryption functions to encrypt financial international data
* Added custom resolver for updateLossPayeeFinancialDetailsInternational
* Updated typedefs
* Added graphql mutuation call to UI
* Added data saving to UI for loss payee financial details international
* Added unit and API tests

### CODE REVIEW

File: src/api/custom-schema/type-defs.ts [hunk 1]
```diff
@@ -248,13 +248,21 @@ const typeDefs = `
     bankAddress: String
   }
 
+  type FinancialInternational {
```
Comment (ttbarnes) [inline src/api/custom-schema/type-defs.ts:251]: `LossPayeeFinancialInternational`?

File: src/api/helpers/map-loss-payee-financial-details-international/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import mapLossPayeeFinancialDetailsInternational from '.';
+import encrypt from '../encrypt';
+import { mockLossPayeeFinancialDetailsInternational } from '../../test-mocks/mock-application';
+import { ApplicationLossPayeeFinancialInternational } from '../../types';
+
+describe('api/helpers/map-loss-payee-financial-details-international', () => {
+  describe('when iban, bicSwiftCode and bankDetails are not defined', () => {
+    const variables = {} as ApplicationLossPayeeFinancialInternational;
+
+    it('should return variables as an empty string or null', () => {
```
Comment (ttbarnes) [inline src/api/helpers/map-loss-payee-financial-details-international/index.test.ts:10]: Can't see anything null here 😛

File: src/ui/server/controllers/insurance/policy/loss-payee-financial-details-international/index.test.ts [hunk 3]
```diff
@@ -144,6 +150,30 @@ describe('controllers/insurance/policy/loss-payee-financial-details-internationa
 
         expect(res.redirect).toHaveBeenCalledWith(expected);
       });
+
+      it('should call mapAndSave.lossPayeeFinancialDetailsInternational once with data from constructPayload function', async () => {
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/policy/loss-payee-financial-details-international/index.test.ts:154]: Do we have tests for `mapAndSave` errors? e.g

```js
describe('when mapAndSave.exportContractAgentServiceCharge does not return a true boolean', () => {
  describe('when mapAndSave.exportContractAgentServiceCharge returns an error', () => {
```

File: src/ui/server/controllers/insurance/policy/map-and-save/loss-payee-financial-details-international/index.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
+import hasFormData from '../../../../../helpers/has-form-data';
+import mapSubmittedData from '../../map-submitted-data/loss-payee-financial-details';
+import save from '../../save-data/loss-payee-financial-details-international';
+import { Application, RequestBody, ValidationErrors } from '../../../../../../types';
+
+/**
+ * lossPayeeFinancialDetailsInternational
+ * mapAndSave lossPayeeFinancialDetails
+ * Map and save any valid nominatedLossPayee fields
+ * @param {Express.Request.body} Express request body
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/policy/map-and-save/loss-payee-financial-details-international/index.ts:10]: Can we rename to this please, trying to align these (only a few left)

```js
* @param {Express.Request.body} Form data
```

File: src/ui/server/controllers/insurance/policy/save-data/loss-payee-financial-details-international/index.test.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  describe('NULL_OR_EMPTY_STRING_FIELDS', () => {
+    it('should have the relevant fieldIds', () => {
+      expect(NULL_OR_EMPTY_STRING_FIELDS).toEqual([IBAN, BIC_SWIFT_CODE, FINANCIAL_ADDRESS]);
+    });
+  });
+
+  describe('when errorList is provided', () => {
+    const mockValidationErrors = generateValidationErrors(FINANCIAL_ADDRESS, 'error', {});
+
+    it(`should call api.keystone.application.update.lossPayeeFinancialDetailsInternational with all fields but not ${FINANCIAL_ADDRESS}`, async () => {
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/policy/save-data/loss-payee-financial-details-international/index.test.ts:33]: Seems the description is invalid, could/should probably be simpler:

```
"should call api.keystone.application.update.lossPayeeFinancialDetailsInternational"
```

File: src/ui/server/controllers/insurance/policy/save-data/loss-payee-financial-details-international/index.test.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+    it('should return the API response', async () => {
+      const result = await save.lossPayeeFinancialDetailsInternational(mockApplication, mockFormBody);
+
+      expect(result).toEqual(mockUpdateApplicationResponse);
+    });
+  });
+
+  describe('when errorList is NOT provided', () => {
+    it('should call api.keystone.application.update.lossPayeeFinancialDetailsInternational with all fields', async () => {
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/policy/save-data/loss-payee-financial-details-international/index.test.ts:52]: Same as above

File: src/api/custom-schema/type-defs.ts (hunk 1 above)
Comment (Zainzzkk) [reply src/api/custom-schema/type-defs.ts]: Cannot be `LossPayeeFinancialInternational` as that is an autogenerated one

**Comment by ttbarnes** [comment]
There's a new SonarCloud issue reported with this line, should be a quick fix

```js
lossPayeeFinancialDetailsInternationalResponse = (await updateLossPayeeFinancialDetailsInternational({}, variables, context)) as SuccessResponse;
```

https://sonarcloud.io/project/issues?resolved=false&pullRequest=2294&id=UK-Export-Finance_exip&open=AY8Gb4CwkFfzVNFJylsN

File: src/api/custom-resolvers/queries/get-application-by-reference-number/index.test.ts [hunk 7]
```diff
@@ -82,7 +82,13 @@ describe('custom-resolvers/get-application-by-reference-number', () => {
...
   describe('when an error occurs', () => {
     it('should throw an error', async () => {
-      await expect(getApplicationByReferenceNumber()).rejects.toThrow('Get application by reference number (GetApplicationByReferenceNumber mutation)');
+      try {
+        await getApplicationByReferenceNumber({}, { referenceNumber: 0 }, context);
+      } catch (err) {
+        const errorString = String(err);
+
+        expect(errorString.includes('Error generating buffer')).toEqual(true);
+      }
```
Comment (ttbarnes) [inline src/api/custom-resolvers/queries/get-application-by-reference-number/index.test.ts:91]: Could this use `rejects.toThrow()`?

File: src/api/helpers/decrypt-financial-uk/index.test.ts [hunk 8]
```diff
@@ -33,23 +33,10 @@ describe('api/helpers/decrypt-financial-uk', () => {
...
-
-      const expected = {
-        sortCode: '',
-        accountNumber: '',
-      };
-
-      expect(result).toEqual(expected);
-    });
-  });
-
```
Comment (ttbarnes) [inline src/api/helpers/decrypt-financial-uk/index.test.ts:48]: What's the reason for removing this test?

File: src/api/helpers/decrypt-financial-uk/index.test.ts [hunk 9]
```diff
@@ -33,23 +33,10 @@ describe('api/helpers/decrypt-financial-uk', () => {
...
   describe('when an error occurs', () => {
     it('should throw an error', async () => {
       try {
-        decryptFinancialUk();
+        decryptFinancialUk({ id: '1', accountNumber: '1', sortCode: '1', accountNumberVector: '1', sortCodeVector: '1' });
```
Comment (ttbarnes) [inline src/api/helpers/decrypt-financial-uk/index.test.ts:39]: Not understanding here how this causes an error?

File: src/api/test-mocks/mock-application.ts [hunk 10]
```diff
@@ -142,13 +142,25 @@ export const mockLossPayeeFinancialDetailsUk = {
   bankAddress: 'Mock UK financial address',
 };
 
+export const mockLossPayeeFinancialDetailsInternational = {
+  iban: 'AAaaa1A1AAaaAAa1AaAaAaaaAaAaaaAaAAAaAaAaAaA=',
+  ibanVector: 'AAaaAAA1aA+1aaa1',
+  bicSwiftCode: 'BBbbb2B1BBbbBBb2BbBbBbbbBbBbbbBbBBBbBbBbBbB',
+  bicSwiftCodeVector: 'BBbbBBB2bB+2bbb2',
```
Comment (ttbarnes) [inline src/api/test-mocks/mock-application.ts:149]: Could we have different values to the UK details?

File: src/api/custom-resolvers/queries/get-application-by-reference-number/index.test.ts (hunk 7 above)
Comment (Zainzzkk) [reply src/api/custom-resolvers/queries/get-application-by-reference-number/index.test.ts:91]: reverted

File: src/api/helpers/decrypt-financial-uk/index.test.ts (hunk 8 above)
Comment (Zainzzkk) [reply src/api/helpers/decrypt-financial-uk/index.test.ts:48]: by accident, readded

**Review by ttbarnes — APPROVED** [review]

File: src/api/test-mocks/mock-application.ts (hunk 10 above)
Comment (Zainzzkk) [reply src/api/test-mocks/mock-application.ts:149]: in next PR

**Pull request merged**